src/stages/stage3_events.c

```diff
@@ -741,939 +81,689 @@ Boss* stage3_spawn_scuttle(cmplx pos) {
+				);
+		play_sound("shot1");
+		WAIT(20);
 	}
 
-	// moonlight rocket rockets
-	int rocket_period = (160 + 20 * (D_Lunatic - global.diff));
+}
 
-	if(!creal(e->args[3]) && !((time + rocket_period/2) % rocket_period)) {
-		Laser *l;
-		float dt = 60;
+TASK(side_swirls_procession, { cmplx start_pos; int count; cmplx p0; cmplx p1; cmplx p2; } ) {
 
-		l = create_lasercurve4c(e->pos, dt, dt, RGBA(1.0, 1.0, 0.5, 0.0), las_sine_expanding, 2.5*dir, M_PI/20, 0.2, 0);
-		PROJECTILE(
-			.proto = pp_ball,
-			.pos = e->pos,
-			.color = RGB(1.0, 0.4, 0.6),
-			.rule = wriggle_rocket_laserbullet,
-			.args = {
-				add_ref(l), dt-1, 1
-			}
-		);
+	int interval = 20;
 
-		l = create_lasercurve4c(e->pos, dt, dt, RGBA(0.5, 1.0, 0.5, 0.0), las_sine_expanding, 2.5*dir, M_PI/20, 0.2, M_PI);
-		PROJECTILE(
-			.proto = pp_ball,
-			.pos = e->pos,
-			.color = RGB(1.0, 0.4, 0.6),
-			.rule = wriggle_rocket_laserbullet,
-			.args = {
-				add_ref(l), dt-1, 1
-			},
-		);
-
-		play_sound("laser1");
+	for(int x = 0; x < ARGS.count; ++x) {
+		cmplx p2 = ARGS.p2;
+		if(!p2) {
+			p2 = 5 + (0.93 + 0.01 * x) * I;
+		}
+		INVOKE_TASK(side_swirl, ARGS.start_pos, ARGS.p0, ARGS.p1, p2);
+		WAIT(interval);
 	}
+}
 
-	// night ignite balls
-	if(creal(e->args[3]) && global.diff > D_Easy) {
-		FROM_TO(300, 1000000, 180) {
-			int cnt = 5, i;
-			for(i = 0; i < cnt; ++i) {
-				PROJECTILE(
-					.proto = pp_ball,
-					.pos = e->pos,
-					.color = RGBA(0.5, 1.0, 0.5, 0),
-					.rule = accelerated,
-					.args = {
-						0, 0.02 * cexp(I*i*2*M_PI/cnt)
-					},
-				);
+// fairies
 
-				if(global.diff > D_Hard) {
-					PROJECTILE(
-						.proto = pp_ball,
-						.pos = e->pos,
-						.color = RGBA(1.0, 1.0, 0.5, 0),
-						.rule = accelerated,
-						.args = {
-							0, 0.01 * cexp(I*i*2*M_PI/cnt)
-						},
-					);
-				}
-			}
+TASK(little_fairy, { cmplx pos; cmplx target_pos; int danmaku_type; int side; }) {
+	// contains stage3_slavefairy and stage3_slavefairy2
+	Enemy *e = TASK_BIND_UNBOXED(create_enemy1c(ARGS.pos, 900, Fairy, NULL, 0));
 
-			// FIXME: better sound
-			play_sound("shot_special1");
-		}
-	}
+	// fade-in
+	e->alpha = 0;
 
-	return 1;
-}
+	INVOKE_TASK_WHEN(&e->events.killed, common_drop_items, &e->pos, {
+			.points = 3,
+			.power = 2,
+			});
 
-void wriggle_moonlight_rocket(Boss *boss, int time) {
-	int i, j, cnt = 1 + global.diff;
-	TIMER(&time)
+	e->move.attraction_point = ARGS.target_pos;
 
-	AT(EVENT_DEATH) {
-		enemy_kill_all(&global.enemies);
-		return;
-	}
+	int shot_interval = 1;
+	int intensity = difficulty_value(90, 70, 50, 30);
 
-	if(time < 0)
-		GO_TO(boss, VIEWPORT_W/2 + VIEWPORT_H*I/2.5, 0.05)
-	else if(time == 0) {
-		for(j = -1; j < 2; j += 2) for(i = 0; i < cnt; ++i)
-			create_enemy3c(boss->pos, ENEMY_IMMUNE, wriggle_slave_visual, wriggle_spell_slave, add_ref(boss), i*2*M_PI/cnt, j);
-	}
-}
+	switch(ARGS.danmaku_type) {
```

Comment:
I really don't like this giant switch. These two should just be separate tasks. You'd have to copy-paste the spawn/escape boilerplate, but it reads a lot nicer that way.